GitHub action

ll-nick opened the pull request:
This adds GitHub Actions to the repository that do the following:

- Run the unit tests on all pushes to any branch
- On a merge to main:
  - Bump the version that's currently defined in the `version` file, add/commit the new version and git tag this commit. By default, this bumps the minor version but this can be adjusted using the #major, #minor, #patch keywords in the commit message.
  - Pulls the newly created tag and builds the deb package and archive releases.
  - Creates a new release on GitHub using the PR description as the release description.

orzechow commented on .github/workflows/bump-version-and-create-release.yaml:
```diff
@@ -0,0 +1,78 @@
+          mkdir -p /tmp/artifacts/
+          docker run --rm -v /tmp/artifacts:/tmp/artifacts util_caching_release cp -r /release /tmp/artifacts/
+
+      - name: Create Release
+        uses: ncipollo/release-action@v1
+        with:
+          artifacts: "/tmp/artifacts/release/*"
+          tag: ${{ needs.bump-version.outputs.new_tag }}
+
```
It would be nice, if the release description would contain the corresponding PR description.
[Currently](https://github.com/KIT-MRT/util_caching/releases), it's the version bump commit message :see_no_evil: 

I imagine this could work:
- save the merge commit message into a tmp file (must be an early github action step, before version bump commit)
- pass this file to the `ncipollo/release-action` action as `bodyFile`

orzechow reviewed (approved):
This is so awesome! :star_struck: 

Two nitpicky points, but feel free to merge.

orzechow reviewed (changes requested):
Sorry, I found a bug:

The latest release v0.1.1 contains the v0.1.0 package :cry: 
https://github.com/KIT-MRT/util_caching/releases/tag/v0.1.1

orzechow commented:
> The readme should now suggest using the deb package as the recommended way of installing it.

Good point!
Feel free to change it with this PR :+1: 

> It would be nice to have multi-architecture builds/releases. If that's not too complicated, I'd integrate that as well.

Well, a header-only release is architecture independent – that's one of it's beauties.
But – I had a wrong CPack variable name which made CPack use the system architecture instead of setting `architecture=all`.
Fixed with #22

> Right now, the default version bump is set to patch. Do you think that's appropriate or should we set it to minor @orzechow ?

Right. `#minor` would be more appropriate.

ll-nick commented:
Ok @orzechow, this one is ready for another review.

 
> >  The readme should now suggest using the deb package as the recommended way of installing it.
> 
> Good point!
> Feel free to change it with this PR 👍

Done in b7d01bec607f1b8097a77a1ad62da8f90dca8917


> > It would be nice to have multi-architecture builds/releases. If that's not too complicated, I'd integrate that as well.
> 
> Well, a header-only release is architecture independent – that's one of it's beauties.
> But – I had a wrong CPack variable name which made CPack use the system architecture instead of setting architecture=all.
> Fixed with https://github.com/KIT-MRT/util_caching/pull/22

Very good point, did not think that through :sweat_smile: 


>  >  Right now, the default version bump is set to patch. Do you think that's appropriate or should we set it to minor @orzechow ?
> 
> Right. #minor would be more appropriate.

Changed the default bump to #minor in c1486dfd8471528061ffeafcbb12363596e43d73

> It would be nice, if the release description would contain the corresponding PR description.

Done in 93676b754abc939fd8a3ad6db852182df4216401. We should be aware that this is [vulnerable to script injections](https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections) but I figured that we would probably notice malicious code in the PR description.

> Sorry, I found a bug:
>
> The latest release v0.1.1 contains the v0.1.0 package 😢

Good catch! This happened because the docker [build-push-action](https://github.com/docker/build-push-action?tab=readme-ov-file#git-context) will use the repository state that caused the workflow, so changes inside the workflow are not considered. This means that the deb package did contain all the changes of the new version except the bumped `version` file which consequently lead to the wrong name of the package.

I fixed it by explicitly checking out the tag we created in the first job in 5adc18d4d23ede5456aca6d471f2fd2d45139c00 and by setting the docker context to `.` in 45d3b44dc25d5ba8df5b117f5e2c84f06321ac37. See [here](https://github.com/docker/build-push-action?tab=readme-ov-file#git-context) for reference on why this is necessary.

ll-nick commented:
Since the PR description will now be the release description too, I updated the original description of the PR to be a little more concise and highlight the contributions of the PR.

orzechow reviewed (approved):
> Ok @orzechow, this one is ready for another review.

Beautiful, thanks a lot! :clap: 

Two suggestions, feel free to merge after considering these.


> > It would be nice, if the release description would contain the corresponding PR description.
> 
> Done in [93676b7](https://github.com/KIT-MRT/util_caching/commit/93676b754abc939fd8a3ad6db852182df4216401). We should be aware that this is [vulnerable to script injections](https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections) but I figured that we would probably notice malicious code in the PR description.

Awesome, the test release v0.2.0 worked out great!

Conserning the vulnerability, I agree that we would probably notice such code.
But, I think we can avoid it altogether, see my code suggestions.


> > Sorry, I found a bug:
> > The latest release v0.1.1 contains the v0.1.0 package 😢
> 
> Good catch! This happened because the docker [build-push-action](https://github.com/docker/build-push-action?tab=readme-ov-file#git-context) will use the repository state that caused the workflow, so changes inside the workflow are not considered. This means that the deb package did contain all the changes of the new version except the bumped `version` file which consequently lead to the wrong name of the package.
> 
> I fixed it by explicitly checking out the tag we created in the first job in [5adc18d](https://github.com/KIT-MRT/util_caching/commit/5adc18d4d23ede5456aca6d471f2fd2d45139c00) and by setting the docker context to `.` in [45d3b44](https://github.com/KIT-MRT/util_caching/commit/45d3b44dc25d5ba8df5b117f5e2c84f06321ac37). See [here](https://github.com/docker/build-push-action?tab=readme-ov-file#git-context) for reference on why this is necessary.

Nice and clean solution! :+1:

ll-nick commented on .github/workflows/bump-version-and-create-release.yaml:
```diff
@@ -0,0 +1,90 @@
+        with:
+          artifacts: "/tmp/artifacts/release/*"
+          tag: ${{ needs.bump-version.outputs.new_tag }}
+          bodyFile: /tmp/pull_request_description.md
```
Ah, yes, much better. It's nice that sometimes things are actually easier than you'd think :joy:

I just tested this and it works just fine :+1:

ll-nick commented on README.md:
````diff
@@ -79,6 +79,15 @@ More usage please check the unittest.
+### Using Debian package (recommended)
+
+We provide a Debian package for easy installation on Debian-based distributions.
+Download the latest `.deb` package from the [releases page](https://github.com/KIT-MRT/util_caching/releases) and install it with `dpkg`:
+
+```bash
+sudo dpkg -i util-caching*.deb
````
Yes, let's do that. I think having a static link here is quite beneficial.

What about the architecture in the file name? As you mentioned, this is header only, so the deb file should work everywhere right?

Should I just set
```
set(CPACK_DEBIAN_FILE_NAME "${PROJECT_NAME}.deb")
```
in `cpack_config.cmake`?
